src/Agent.Listener/Configuration/LinuxServiceControlManager.cs

```diff
@@ -1,14 +1,281 @@
+        }
+
+        protected virtual bool CheckIfSystemdExists()
+        {
+            Trace.Info(nameof(CheckIfSystemdExists));
+            try
+            {
+                var commName = File.ReadAllText(InitFileCommandLocation).Trim();
+                return commName.Equals("systemd", StringComparison.OrdinalIgnoreCase);
+            }
+            catch (Exception ex)
+            {
+                Trace.Error(ex.ToString());
+                _term.WriteError(StringUtil.Loc("CanNotFindSystemd"));
+
+                return false;
+            }
+        }
+
+        protected virtual bool CheckServiceExists(string serviceName)
+        {
+            Trace.Info(nameof(CheckServiceExists));
+            if (string.IsNullOrEmpty(serviceName))
+            {
+                throw new ArgumentNullException("serviceName");
+            }
+
+            try
+            {
+                var unitFile = new FileInfo(Path.Combine(SystemdPathPrefix, serviceName));
+                return unitFile.Exists;
+            }
+            catch (Exception ex)
+            {
+                Trace.Error(ex);
+
+                // If we can check if the service exists we can't configure either. We can't ignore this error.
+                throw;
+            }
+        }
+
+        private void ChangeOwnershipToLoginUser(Dictionary<string, int> filesToChange)
+        {
+            // Before starting the service chown/chmod the _diag and settings files to the current user.
+            // Since we started with sudo, the _diag will be owned by root. Change this to current login user
+
+            Trace.Info(nameof(ChangeOwnershipToLoginUser));
+
+            try
+            {
+                StringBuilder loginUser = new StringBuilder();
+                getlogin_r(loginUser, MaxUserNameLength);
+                string userName = loginUser.ToString();
+
+                IntPtr statPtr = getpwnam(userName);
+                var stat = Marshal.PtrToStructure<LoginStat>(statPtr);
+                foreach (var file in filesToChange)
+                {
+                    Trace.Info("Changing ownership of {0} to current user {1}", file.Key, userName);
+                    chown(file.Key, stat.pw_uid, stat.pw_gid);
+
+                    Trace.Info("Changing permission of {0} to {1}", file.Key, file.Value);
+                    chmod(file.Key, file.Value);
+                }
+            }
+            catch (Exception ex)
+            {
+                Trace.Error(ex);
+                _term.WriteError(StringUtil.Loc("CannotChangeOwnership"));
+            }
+        }
+
+        private void EnableService(string serviceName)
+        {
+            Trace.Info(nameof(EnableService));
+            try
+            {
+                ExecuteSystemdCommand("enable " + serviceName);
+            }
+            catch (Exception)
+            {
+                _term.WriteError(StringUtil.Loc("LinuxServiceStartFailed"));
```

Comment:
I was thinking not able to start the service can be a warning though.. changed it throw error anyway..